Specifications/QIR/Reserved-Operations.md

```diff
@@ -0,0 +1,41 @@
+# Reserved Operation Names
+
+As mentioned [here](Quantum-Runtime.md), we do reserve a set of quantum operation names.
+While targets are not required to implement any or all of these operations,
+if they define an operation with a reserved name that operation's effect is required
+to match the specified action.
+This ensures that quantum code that uses, say, a `CNOT` operation can rely on the
+effect of that operation being consistent across targets that define it.
+
+In this table, operation names should be considered case-insensitive.
+That is, if a target implements `RX` instead of `Rx`, its effect still must match that
+defined in the table.
+
+The matrices in this tables below take some time to load.
+Please be patient!
+
+| Operation Name | QIR Implementation | Description | Notes |
+|----------------|-----------------|-------------|-------|
+| H | [`__quantum__qis__h__body`](https://github.com/microsoft/qsharp-runtime/blob/5a379fd1c72ea19cb50d6acc501c6896894e96ac/src/Qir/Runtime/lib/QSharpCore/qsharp-core-qis.ll#L127) | The [Hadamard operation](https://docs.microsoft.com/en-us/qsharp/api/qsharp/microsoft.quantum.intrinsic.h) | [Single-qubit operations](https://docs.microsoft.com/en-us/azure/quantum/concepts-the-qubit#single-qubit-operations), [Other Single-Qubit Cliffords](https://docs.microsoft.com/en-us/azure/quantum/user-guide/libraries/standard/prelude#other-single-qubit-cliffords) |
+| T | [`__quantum__qis__t__body`](https://github.com/microsoft/qsharp-runtime/blob/5a379fd1c72ea19cb50d6acc501c6896894e96ac/src/Qir/Runtime/lib/QSharpCore/qsharp-core-qis.ll#L224) | The [T operation](https://docs.microsoft.com/en-us/qsharp/api/qsharp/microsoft.quantum.intrinsic.t) | [Single-qubit operations](https://docs.microsoft.com/en-us/azure/quantum/concepts-the-qubit#single-qubit-operations), [Rotations](https://docs.microsoft.com/en-us/azure/quantum/user-guide/libraries/standard/prelude#rotations). ![T = \begin{bmatrix} 1 & 0 \\ 0 & e^{i\pi/4} \\ \end{bmatrix} = \begin{bmatrix} 1 & 0 \\ 0 & \frac{1}{\sqrt{2}}+ \frac{1}{\sqrt{2}}i \\ \end{bmatrix}](https://render.githubusercontent.com/render/math?math=%5Cdisplaystyle+T+%3D+%5Cbegin%7Bbmatrix%7D+1+%26+0+%5C%5C+0+%26+e%5E%7Bi%5Cpi%2F4%7D+%5C%5C+%5Cend%7Bbmatrix%7D+%3D+%0A%5Cbegin%7Bbmatrix%7D+1+%26+0+%5C%5C+0+%26+%5Cfrac%7B1%7D%7B%5Csqrt%7B2%7D%7D%2B+%5Cfrac%7B1%7D%7B%5Csqrt%7B2%7D%7Di+%5C%5C+%5Cend%7Bbmatrix%7D) |
+
+| Operation Name | QIR Implementation | Description | Notes |
```

Comment:
I would merge the two tables.

Reply:
I can do that, it is easy. 

I just wanted to emphasize that the H and T operations cannot be implemented with other operations. But the others can be implemented based on H and T (but I'm not sure if the rotations can be implemented with H and T). 

When recently I needed to study the fundamentals of the quantum computing, it was curious for me to realize that a number of operations can be implemented based on a very small set of the other basic operations. And no any piece of information was providing an explicit picture/table of such a layered structure of commands. I just tried to make a step in that direction, to give a hint to the new hardware vendors, that their implementation can be simplified if they implement the quantum operation in a certain order.

For the same purposes I added the formulas showing the dependencies between the operations, like `Z = T^4 = S^2 = HXH`.

Reply:
Even though that is true mathematically -- and is true for some other sets of operations as well, not just H and T -- it's not how most actual hardware works. The simplest, easiest to implement operations vary wildly depending on the type of hardware, and so hardware vendors are constrained by that rather than by the algebraic structure of the operations.

I would have one table, sorted alphabetically.

Reply:
> I just wanted to emphasize that the H and T operations cannot be implemented with other operations. But the others can be implemented based on H and T (but I'm not sure if the rotations can be implemented with H and T).

The `H`,  `S`, `CNOT`, and `T` operations are universal for compiling arbitrary precision implementations of arbitrary operations. That may not be the particular decomposition used by a particular implementer, however — some targets may validly treat `H` as derived from other operations unique to that target, or may choose another non-Clifford operation such as `V` instead of `T` to build universal gate sets.

Reply:
Thank you for explaining, Alan and Chris. That is very good to know :-). 
I'll merge the tables and sort alphabetically.